### PR TITLE
Make TransformationRegistry immutable

### LINKED ISSUE #119: Make TransformationRegistry immutable

### PR DESCRIPTION
**Pull request by broccolai**
Closes #119

### CODE REVIEW

File: src/main/java/net/kyori/adventure/text/minimessage/MiniMessage.java [hunk 1]
```diff
@@ -229,31 +190,13 @@ static Builder builder() {
...
+     * Uses the supplied transformation registry.
      *
-     * @param types the types of transformations to add
+     * @param transformationRegistry the transformation registry to use
      * @return this builder
      * @since 4.1.0
      */
-    @SuppressWarnings("unchecked")
-    @NotNull Builder transformations(final TransformationType<? extends Transformation>... types);
+    @NotNull Builder transformationRegistry(final TransformationRegistry transformationRegistry);
```
Comment (kashike) [inline src/main/java/net/kyori/adventure/text/minimessage/MiniMessage.java:199]:
```suggestion
    @NotNull Builder transformations(final TransformationRegistry transformationRegistry);
```

File: src/main/java/net/kyori/adventure/text/minimessage/transformation/TransformationRegistry.java [hunk 2]
```diff
@@ -23,146 +23,110 @@
...
-  public static final TransformationRegistry EMPTY = new TransformationRegistry();
-
-  static {
-    EMPTY.clear();
-  }
-
-  private final List<TransformationType<? extends Transformation>> types = new ArrayList<>();
+public interface TransformationRegistry {
+  TransformationRegistry EMPTY = new TransformationRegistryImpl(Collections.emptyList());
+  TransformationRegistry STANDARD = TransformationRegistry.builder().build();
```
Comment (kashike) [inline src/main/java/net/kyori/adventure/text/minimessage/transformation/TransformationRegistry.java:45]: Move these to the impl class.

File: src/main/java/net/kyori/adventure/text/minimessage/transformation/TransformationRegistry.java [hunk 3]
```diff
@@ -23,146 +23,110 @@
...
+   * @return whether any transformation exists
    * @since 4.1.0
    */
-  @SafeVarargs
-  public TransformationRegistry(final TransformationType<? extends Transformation>... types) {
-    for (final TransformationType<? extends Transformation> type : types) {
-      this.register(type);
-    }
-  }
+  boolean exists(String name, Function<String, ComponentLike> placeholderResolver);
```
Comment (kashike) [inline src/main/java/net/kyori/adventure/text/minimessage/transformation/TransformationRegistry.java:68]: make variables final

File: src/main/java/net/kyori/adventure/text/minimessage/transformation/TransformationRegistry.java [hunk 4]
```diff
@@ -23,146 +23,110 @@
...
-    this.register(TransformationType.HOVER_EVENT);
-    this.register(TransformationType.CLICK_EVENT);
-    this.register(TransformationType.KEYBIND);
-    this.register(TransformationType.TRANSLATABLE);
-    this.register(TransformationType.INSERTION);
-    this.register(TransformationType.FONT);
-    this.register(TransformationType.GRADIENT);
-    this.register(TransformationType.RAINBOW);
-  }
+  @Nullable Transformation get(String name, List<TagPart> inners, Map<String, Template> templates, Function<String, ComponentLike> placeholderResolver, Context context);
```
Comment (kashike) [inline src/main/java/net/kyori/adventure/text/minimessage/transformation/TransformationRegistry.java:58]: make variables final

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by MiniDigger** [comment]
thanks!

**Pull request merged**